config: add ordered dictionary (internal module)

Totktonada opened the pull request:
In development of the configuration module we met several cases, when an order of appearance of some items is as important as a constant time access using a key.

This commit adds a collection that serves such a need. It is to be used internally in the src/box/lua/config code.

The module is inspired by Python's collections.OrderedDict. See the description in the module code for details.

Totktonada commented:
`checkpatch.pl` complains about using `od` as a variable name in Lua (it states that it is misspelled 'of'). Filed https://github.com/tarantool/checkpatch/issues/70.

Lord-KA commented on src/box/lua/config/utils/odict.lua:
```diff
@@ -0,0 +1,209 @@
+end
+
+-- Delete the key from the id<->key mappings.
+local function trash(ctx, key)
```
I would prefer a different function name here, because "to trash something" has a different meaning. Maybe: `remove_key(ctx, key)` or `drop_key(ctx, key)`

Totktonada replied:
I've started from `set` and `drop` functions, but than found that I need functions to control tracking of the keys separately from assignments into the key-value table.

The idea of the `trash` and `track` names are that they have no association with the key-value table and dedicated only to the internal tracking structures. IOW, they have no `drop`/`delete`/`remove` and `set`/`add`/`append` words. Also, `trash` reminds the `TRASH` macro from `src/trivia/util.h`.

I like these names, but I would hear thoughts from other reviewers before a decision to keep or change them.

Totktonada replied:
Renamed.

```diff
diff --git a/src/box/lua/config/utils/odict.lua b/src/box/lua/config/utils/odict.lua
index dce3e9ad52..ffefb00ec0 100644
--- a/src/box/lua/config/utils/odict.lua
+++ b/src/box/lua/config/utils/odict.lua
@@ -108,7 +108,7 @@ local function _pairs(od)
 end
 
 -- Delete the key from the id<->key mappings.
-local function trash(ctx, key)
+local function release(ctx, key)
     local id = ctx.key2id[key]
     if id ~= nil then
         ctx.key2id[key] = nil
@@ -149,7 +149,7 @@ local function reindex(od, ctx)
         local key = ctx.id2key[id]
         -- Drop the given key-id pair from the key<->id mappings.
         if key ~= nil then
-            trash(ctx, key)
+            release(ctx, key)
         end
         -- Add the new key-id pair into the key<->id mappings.
         if key ~= nil and od[key] ~= nil then
@@ -168,7 +168,7 @@ local mt = {
 
         -- If a field with this key had existed, we should update
         -- the key<->id mappings to let gen() know the new order.
-        trash(ctx, k)
+        release(ctx, k)
 
         -- If a non-existing field is assigned to nil, we have
         -- nothing to do.
```

Lord-KA commented on src/box/lua/config/utils/odict.lua:
```diff
@@ -0,0 +1,209 @@
+        ctx.key2id[key] = nil
+        ctx.id2key[id] = nil
+
+        -- No need to grow the IDs if the same key is assigned and
+        -- deleted in a loop.
+        --
+        -- for <...> do
+        --     od.x = 'x'
+        --     od.x = nil
+        -- end
```
Do you mean that we should have sequential ids?

Totktonada replied:
Internal IDs? Nope.

This code may be removed without any visible behavior change except more frequent reindexing in cases like the quoted one.

Lord-KA commented on src/box/lua/config/utils/odict.lua:
```diff
@@ -0,0 +1,209 @@
+end
+
+-- Track the new key in the id<->key mappings.
+local function track(ctx, key)
+    local id = ctx.max_id + 1
+    ctx.id2key[id] = key
+    ctx.key2id[key] = id
+    ctx.max_id = id
+end
```
Nit:
This seems more like a `new_index()`/`add_key()` function.
Actually, it would be nice to have `drop_key()` and `add_key()` instead of `trash()` and `track()`.

Totktonada replied:
Answered [above](https://github.com/tarantool/tarantool/pull/9730#discussion_r1505606234).

Totktonada replied:
Updated, see the linked discussion thread.

Lord-KA commented on src/box/lua/config/utils/odict.lua:
```diff
@@ -0,0 +1,209 @@
+            trash(ctx, key)
+        end
+        -- Add the new key-id pair into the key<->id mappings.
+        if key ~= nil and od[key] ~= nil then
```
Nit:
I may be mistaken, but it seems like we don't need `key ~= nil` check, do we?

Totktonada replied:
Yep. We can use `nil` as a table index key and will always get `nil` as a result (but we can't assign a `nil` key). However, I like this condition to be written explicitly, because:

1. There is the `key ~= nil` condition above and it is easier to spot that both `trash()` and `track()` are skipped in case of `key == nil`.
2. Using `nil` as an index key for a table is a trick (not known by any Lua user) and it anyway should be commented. An explicit check looks better for me than an implicit one + comment.

Lord-KA replied:
If you prefer so, sure.

Lord-KA reviewed (approved):
Thanks for the patch. Overall looks good, see my 5 suggestions below.

Totktonada commented on src/box/lua/config/utils/odict.lua:
```diff
@@ -0,0 +1,209 @@
+--  | -- print: b, 2
+--  | -- print: c, 3
+--
+-- If an element is deleted and added again, it is added to the
+-- end.
+--
+--  | local od = odict.new()
+--  |
+--  | od.a = 1
+--  | od.b = 2
+--  | od.c = 3
+--  |
+--  | od.b = nil
+--  | od.b = 4
+--  |
+--  | for k, v in odict.pairs(od) do
+--  |     print(k, v)
+--  | end
+--  | -- print: a, 1
+--  | -- print: c, 3
+--  | -- print: b, 4
```
Thanks for noticing! I've filed #9747. I'll also mention this detail in the module header comment.

I'm against implicit transformation from a mapping to an array. I would rather leave the ordered serialization unimplemented (but keep the shape of the collection as is) and implement such serialization mechanism later (without changing the shape).

Totktonada replied:
While we're in a discussion, I've added a note.

```diff
diff --git a/src/box/lua/config/utils/odict.lua b/src/box/lua/config/utils/odict.lua
index 925e2d1f43..dce3e9ad52 100644
--- a/src/box/lua/config/utils/odict.lua
+++ b/src/box/lua/config/utils/odict.lua
@@ -48,6 +48,10 @@
 --
 -- If an element is changed (without prior deletion), it remains
 -- on the same position.
 --
+-- Beware: console shows the fields as unordered. The same for the
+-- serialization into JSON/YAML/MessagePack formats. It should be
+-- solved after gh-9747.
 
 local REINDEX_THRESHOLD = 1000
```

igormunkin replied:
Well, the idea is quite nice, but I agree with @Totktonada's argument from our offline discussion: though there is already the similar prisedanie with `box.cfg`, I doubt we have to manually patch `pairs` builtin for this. The reason this hack is implemented for `box.cfg` relates to this object nature: it's the singleton being often observed by humans. But do we need implement this hack for such a basic thing? I don't think so.

igormunkin replied:
The same argument works for `__serialize`, `compat` table and odict object.

Totktonada replied:
@Lord-KA Can I resolve this discussion leaving the 'beware' comment in the code and the issue to return here back later?

Lord-KA replied:
I am okay with solving it separately in https://github.com/tarantool/tarantool/issues/9747.

ImeevMA reviewed (approved):
Thank you for the patch! LGTM.

Totktonada commented:
> `checkpatch.pl` complains about using `od` as a variable name in Lua (it states that it is misspelled 'of'). Filed https://github.com/tarantool/checkpatch/issues/70.

Fixed in PR #9752. Rebased this patch on top of it.
